### PR TITLE
bpo-46417: Fix _PyStaticType_Dealloc()

### PR DESCRIPTION
**Pull request by vstinner**
_PyStaticType_Dealloc() now only calls PyObject_ClearWeakRefs() if
the call is not going to fail.

https://bugs.python.org/issue46417

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -4094,19 +4092,29 @@ _PyStaticType_Dealloc(PyTypeObject *type)
     Py_CLEAR(type->tp_cache);
     // type->tp_subclasses is NULL
 
+    // PyObject_ClearWeakRefs() raises an exceptions if Py_REFCNT() != 0
```
Comment (thatbirdguythatuknownot) [inline Objects/typeobject.c:4095]: Grammar error/typo in the comments.

Comment (thatbirdguythatuknownot) [inline Objects/typeobject.c]: Grammar error/typo in the comment.

File: Objects/typeobject.c [hunk 2]
```diff
@@ -4094,19 +4092,29 @@ _PyStaticType_Dealloc(PyTypeObject *type)
...
 {
-    /* Assert this is a heap-allocated type object */
+    // Assert this is a heap-allocated type object
     _PyObject_ASSERT((PyObject *)type, type->tp_flags & Py_TPFLAGS_HEAPTYPE);
+
     _PyObject_GC_UNTRACK(type);
 
     type_dealloc_common(type);
 
+    // PyObject_ClearWeakRefs() raises an exceptions if Py_REFCNT() != 0
```
Comment (thatbirdguythatuknownot) [inline Objects/typeobject.c]: Grammar error/typo in the comment again.

File: Objects/typeobject.c (hunk 1 above)
Comment (vstinner) [reply Objects/typeobject.c]: "an exceptions", oh I see. I will fix it later. First, I need to merge https://github.com/python/cpython/pull/30809 and then rebase this PR on top of it.

**Pull request merged**